### DISCUSSION TITLE
ORM returning dictionary

### DISCUSSION THREAD

**Discussion by cpdivers** [sqlalchemy/sqlalchemy#7926] [Usage Questions] [answered]
Hello everyone,

It should be a simple question

SQLAlchemy Raw, this is working great, and I convert the result into a Dictionary,

```
query = "SELECT * from set_and_minifig_inventory where SETNUMBER= :set_id"
results = engine.execute(text(query),{'set_id':set_id})
allresults = [dict(row) for row in results.fetchall()]
for p in allparts:
  print(p['id']) ---> Works great
```

Is there a way to return a dictionary with ORM ? if yes how ?
This is what I have:
```
allparts = SetAndMinifigInventory.query.filter_by(SETNUMBER = set_id).all()
for p in allparts:
    #print(p['id']) <---------------- THIS FAILS
    print(p) --------------------> RETURNS <SetAndMinifigInventory 753881> which is from my POV ugly
```

**Comment by zzzeek** [comment, marked as answer]
Using SQLAlchemy 1.4, the Query and Session.execute() returns Row objects (tuples) when you select individual columns (`session.query(Model.id, Model.name, Model, ...)` or tables (`session.query(Model.__table__)`).  Row can be turned into a dictionary using ._mapping.

SQLAlchemy 1.3 and earlier, the rows returned by Session.query() can be turned into dictionaries using dict(row).

https://docs.sqlalchemy.org/en/14/orm/queryguide.html#selecting-orm-entities-and-attributes

**Reply by cpdivers** [reply]
```
    print("=================================================")
    separator(f"{whoami()}")
    print("=================================================")
    allparts = db.session.query(UBuildlistPart.__table__).all()
    print(allparts)
    print("=================================================")
    allparts2 = [dict(row) for row in allparts]
    print(allparts2)
```
print(allparts) returns:
I see no dictionary with the column name in here.
```
(1, 9, '4592c05', 11, 2, None, None, None, None), (2, 9, '30374', 11, 4, None, None, None, None), (3, 9, '30359b', 85, 2, None, None, None, None), (4, 9, '6020', 11, 1, None, None, None, None), (5, 9, '2436', 86, 1, None, None, None, None), (6, 9, '3005', 1, 4, None, None, None, None), (7, 9, '3004', 59, 3
```

but if I do add:
```
allparts2 = [dict(row) for row in allparts]
```
I get what I want:
```
[{'id': 1, 'buildlist_id': 9, 'part_id': '4592c05', 'color_id': 11, 'qty': 2, 'taken_from_inventory_qty': None, 'added_to_wishlist_qty': None, 'taken_date': None, 'added_date': None}, {'id': 2, 'buildlist_id': 9, 'part_id': '30374', 'color_id': 11, 'qty': 4, 'taken_from_inventory_qty': None
```



It seems I have answered my own question !

**Reply by kiddten** [reply]
how to convert `Row` object back to ORM object correctly?

**Reply by zzzeek** [reply]
it depends on if the Row has an ORM object inside of it already, just extract it, or if it has individual columns, I would change the query being executed so that it returns ORM objects instead of columns.

**Reply by kiwavi** [reply]
_mapping method works well except when the row data is an array. In my case, it converts all rows to dictionaries except for the row that's a json_agg output. It skips it. What could be the cause? 
PS: the function  that outputs the array is func.array_agg(models.Orders.__table__.table_valued())

Interestingly, when I run ._fields method on the row it's not among the fields

**Reply by CaselIT** [reply]
The issue is likely that no name was given to it. Add a label to that function and you should be able to find it among fields and the dict keys

**Reply by kiwavi** [reply]
That works. Thanks

**Comment by nanarino** [comment]
SQLAlchemy 1.4
 I can't seem to get the `dict` and `_mapping`  to work 🤔🤔 ，So I use the mixin ：

```python
from sqlalchemy.orm import registry
from sqlalchemy import Table
from typing import NewType, Optional, Iterable, Union

mapper_registry = registry()
Base = mapper_registry.generate_base()
Model = NewType('Model', Base)
metadata = Base.metadata

def table(mapper: Union[Model, type[Model]]) -> Table:
    return mapper.__table__

class dict_able_mixin():
    def keys(self) -> Iterable:
        if isinstance(self, Base):
            return map(lambda c: c.key, table(self).columns)
        else:
            return []
    def __getitem__(self, key: str):
        return getattr(self, key)
```


```python
class Demo(dict_able_mixin, Base):
    __tablename__ = 'demo'
    id = Column(Integer, primary_key=True)
    name = Column(String(255))

async with session.begin():
    _orm = select(Demo).where(Demo.id == 1)
    result: Demo = (await session.execute(_orm)).scalars().first()
    print(result) # <Demo object at 0x000001506DCF2E50>
    print(dict(result)) # {'id': 1, 'name': 'test1'}
    print(result._mapping) # 'Demo' object has no attribute '_mapping'
```

**Reply by CaselIT** [reply]
Hi

if you use `scalars` you don't get rows back, but the first column only as a list of scalar values. If you remove it, you will get back a row object, that contains a single column, with the instances of Demo.

If you instead want the single column of Demo you should run that statement using a connection (obtained using `session.connection()`). This will give you row objects with two columns, one of `id` and one for `name`
